### PR TITLE
perf: Add react-query, switch to react query for groups and hark state management.

### PR DESCRIPTION
**Pull request by patosullivan**
This PR adds react query and persists our react-query data to indexedDB. We're starting with groups and hark state data for now. Zustand has been replaced with react-query for all groups and hark related state syncing and cacheing.

### CODE REVIEW

**Comment by patosullivan** [comment]
Ok, I think this is ready to go @arthyn

**Comment by arthyn** [comment]
cool I'm about halfway through

File: ui/src/app.tsx [hunk 1]
```diff
@@ -282,8 +293,9 @@ function ActivityRoute({ isInGroups = false }: { isInGroups: boolean }) {
 }
 
 function GroupsRoutes({ state, location, isMobile, isSmall }: RoutesProps) {
-  const groups = useGroups();
-  const isInGroups = _.isEmpty(groups) ? false : true;
+  const groups = queryClient.getQueryCache().find(['groups'])?.state.data;
+  const isInGroups =
+    groups !== undefined ? (_.isEmpty(groups) ? false : true) : true;
```
Comment (arthyn) [inline ui/src/app.tsx:298]:
```suggestion
    groups !== undefined ? !_.isEmpty(groups) : true;
```

**Comment by patosullivan** [comment]
Looking into an issue with the "Mark All as Read" button rn.

File: ui/src/state/groups/groups.ts [hunk 2]
```diff
@@ -47,1094 +40,1227 @@ function groupAction(flag: string, diff: GroupDiff) {
...
-    if (disableLoading || (preview && !shouldLoad(nest))) return;
+  if (rest.isLoading) {
+    return null;
+  }
 
-    newAttempt(nest);
-    getChannelPreview();
-  }, [getChannelPreview, preview, nest, disableLoading]);
+  if (rest.isError) {
+    throw new Error('Failed to fetch group index');
```
Comment (arthyn) [inline ui/src/state/groups/groups.ts:1195]: when used like a normal hook this means if it fails we'll hit the error boundary and they'll have to refresh

**Review by arthyn — CHANGES_REQUESTED** [review]
There's a few issues here that we should probably address. Other things to consider are more big picture like making subscription tracking have some default so we don't have to always have the giant check and also giving all our queries initial data so that we don't have to do as much "safe" dereferencing.

File: ui/src/components/References/GroupReference.tsx [hunk 3]
```diff
@@ -29,21 +29,16 @@ function GroupReference({
   description,
 }: GroupReferenceProps) {
   const gang = useGang(flag);
+  const preview = useGangPreview(flag, isScrolling);
```
[the remark this answers is not in the document]
Comment (arthyn) [reply ui/src/components/References/GroupReference.tsx:32]: maybe not, the way you handled here is fine

File: ui/src/app.tsx [hunk 4]
```diff
@@ -635,10 +648,18 @@ function RoutedApp() {
...
           <meta name="theme-color" content={userThemeColor} />
         </Helmet>
-        <TooltipProvider skipDelayDuration={400}>
-          <App />
-          <Scheduler />
-        </TooltipProvider>
+        <PersistQueryClientProvider
+          client={queryClient}
+          persistOptions={{
+            persister: indexedDBPersistor(`${window.our}-${app}`),
```
[the remark this answers is not in the document]
Comment (patosullivan) [reply ui/src/app.tsx:653]: good idea

File: ui/src/state/hark.ts [hunk 5]
```diff
@@ -35,106 +21,132 @@ function harkAction(action: HarkAction) {
...
           resolve();
         },
       });
-    }),
-  sawSeam: async (seam) =>
+    });
+
+  return useMutation(mutationFn, {
+    onMutate: async (variables) => {
+      await queryClient.cancelQueries(['carpet', variables.rope.group]);
```
[the remark this answers is not in the document]
Comment (patosullivan) [reply ui/src/state/hark.ts]: fixed in my latest commit

File: ui/src/groups/GroupAdmin/GroupInfoEditor.tsx [hunk 6]
```diff
@@ -65,55 +66,69 @@ export default function GroupInfoEditor({ title }: ViewProps) {
...
-            .getState()
-            .setSecret(groupFlag, values.privacy === 'secret');
+          });
+
+          setSecretMutation({
+            flag: groupFlag,
+            isSecret: values.privacy === 'secret',
+          });
+        }
+        if (editStatus === 'success' || privacyChanged) {
```
[the remark this answers is not in the document]
Comment (patosullivan) [reply ui/src/groups/GroupAdmin/GroupInfoEditor.tsx:112]: Will look into it.

File: ui/src/groups/FindGroups.tsx [hunk 7]
```diff
@@ -119,10 +105,7 @@ export default function FindGroups({ title }: ViewProps) {
        * for peers where a route has to be established this can take
        * upwards of thirty seconds.
        */
-      const results: GroupIndex = await asyncCallWithTimeout(
-        useGroupState.getState().index(preSig(ship)),
-        30 * 1000
-      );
+      const results: GroupIndex | null = data;
```
[the remark this answers is not in the document]
Comment (patosullivan) [reply ui/src/groups/FindGroups.tsx:108]: yeah, idk why this is there. might have had some idea and then backed out of it, didn't clean this part up.

File: ui/src/groups/GroupAdmin/GroupInfoEditor.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (patosullivan) [reply ui/src/groups/GroupAdmin/GroupInfoEditor.tsx:112]: Passed it to the onSuccess callback.

File: ui/src/logic/useReactQuerySubscription.tsx [hunk 8]
```diff
@@ -40,13 +40,17 @@ export default function useReactQuerySubscription({
...
     );
 
-    useSubscriptionState.getState().subscribe({
-      app,
-      path,
-      event: () => {
-        queryClient.invalidateQueries(queryKey);
-      },
-    });
+    await useSchedulerStore.getState().wait(
```
Comment (arthyn) [inline ui/src/logic/useReactQuerySubscription.tsx:43]:
```suggestion
    useSchedulerStore.getState().wait(
```

File: ui/src/logic/useReactQuerySubscribeOnce.tsx [hunk 9]
```diff
@@ -16,15 +16,44 @@ export default function useReactQuerySubscribeOnce<T>({
...
-        event: resolve,
-      });
+  const fetchData = async () => {
+    let timeoutId: NodeJS.Timeout | undefined;
+    let subscriptionId: number | undefined;
+
+    const timeoutPromise = new Promise<never>((_, reject) => {
+      timeoutId = setTimeout(() => {
+        reject(new Error('subscribeOnce timed out'));
+      }, 5000);
```
Comment (arthyn) [inline ui/src/logic/useReactQuerySubscribeOnce.tsx:26]: May want to open this up as a parameter

**Review by arthyn — APPROVED** [review]
lgtm, just the two nits then gtg

File: ui/src/logic/useReactQuerySubscribeOnce.tsx (hunk 9 above)
Comment (patosullivan) [reply ui/src/logic/useReactQuerySubscribeOnce.tsx:26]: done

**Pull request merged**